File: src/android/ResultHandler.java
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.trinity.plugins.hive;

import android.telecom.Call;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import org.elastos.hive.*;
import org.elastos.hive.Void;
import org.elastos.hive.File;

import java.nio.ByteBuffer;

class ResultHandler<T extends Result> implements Callback<T> {
    private final int handlerId;
    private final CallbackContext callbackContext;
    private final ByteBuffer byteBuffer;

    ResultHandler(int id, CallbackContext callbackContext) {
        this.handlerId = id;
        this.callbackContext = callbackContext;
        this.byteBuffer = null;
    }

    ResultHandler(int id, CallbackContext callbackContext, ByteBuffer buffer) {
        this.handlerId = id;
        this.callbackContext = callbackContext;
        this.byteBuffer = buffer;
    }

    private void sendEvent(JSONObject info) throws JSONException {
        if (callbackContext != null) {
            info.put("hid", handlerId);
            PluginResult res = new PluginResult(PluginResult.Status.OK, info);
            res.setKeepCallback(true);
            callbackContext.sendPluginResult(res);
        }
    }

    @Override
    public void onError(HiveException ex) {
        try {
            JSONObject ret = new JSONObject();
            ret.put("error", ex.getMessage());
            sendEvent(ret);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onSuccess(T body) {
        JSONObject ret;
        try {
            if (body instanceof Drive) {
                ret = driveToJson((Drive)body);
            } else if (body instanceof Directory) {
                ret = directoryToJson((Directory)body);
            } else if (body instanceof File) {
                ret = fileToJson((File)body);
            } else if (body instanceof Client.Info) {
                ret = clientInfoToJson((Client.Info)body);
            } else if (body instanceof Drive.Info) {
                ret = driveInfoToJson((Drive.Info)body);
            } else if (body instanceof Directory.Info) {
                ret = dirInfoToJson((Directory.Info)body);
            } else if (body instanceof File.Info) {
                ret = fileInfoToJson((File.Info)body);
            } else if (body instanceof ItemInfo) {
                ret = itemInfoToJson((ItemInfo)body);
            } else if (body instanceof Children) {
                ret = childrenToJson((Children) body);
            } else if (body instanceof Length) {
                ret = lengthToJson((Length)body);
            } else {
                ret = hiveVoidToJson((Void)body);
            }

            if (ret != null) sendEvent(ret);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private JSONObject driveToJson(Drive drive) throws JSONException {
        ObjectMap  map = ObjectMap.acquire(ObjectMap.DRIVE);
        Integer  objId = System.identityHashCode(drive);
        JSONObject ret = new JSONObject();

        ObjectMap.toDriveMap(map).put(objId, drive);
        ret.put("id", objId);
        return ret;
    }

    private JSONObject directoryToJson(Directory directory) throws JSONException {
        ObjectMap  map = ObjectMap.acquire(ObjectMap.DIR);
        Integer  objId = System.identityHashCode(directory);
        JSONObject ret = new JSONObject();

        ObjectMap.toDirMap(map).put(objId, directory);
        ret.put("id", objId);
        return ret;
    }

    private JSONObject fileToJson(File file) throws JSONException {
        ObjectMap  map = ObjectMap.acquire(ObjectMap.FILE);
        Integer  objId = System.identityHashCode(file);
        JSONObject ret = new JSONObject();

        ObjectMap.toFileMap(map).put(objId, file);
        ret.put("id", objId);
        return ret;
    }

    private JSONObject clientInfoToJson(Client.Info info) throws JSONException {
        JSONObjectHolder<Client.Info> holder;

        holder = new JSONObjectHolder<Client.Info>(info);
        holder.put(Client.Info.userId)
              .put(Client.Info.name)
              .put(Client.Info.email)
              .put(Client.Info.phoneNo)
              .put(Client.Info.region);

        return holder.get();
    }

    private JSONObject driveInfoToJson(Drive.Info info) throws JSONException {
        JSONObjectHolder<Drive.Info> holder;

        holder = new JSONObjectHolder<Drive.Info>(info);
        holder.put(Drive.Info.driveId);

        return holder.get();
    }

    private JSONObject dirInfoToJson(Directory.Info info) throws JSONException {
        JSONObjectHolder<Directory.Info> holder;

        holder = new JSONObjectHolder<Directory.Info>(info);
        holder.put(Directory.Info.itemId)
              .put(Directory.Info.name)
              .put(Directory.Info.childCount);

        return holder.get();
    }

    private JSONObject fileInfoToJson(File.Info info) throws JSONException {
        JSONObjectHolder<File.Info> holder;

        holder = new JSONObjectHolder<File.Info>(info);
        holder.put(File.Info.itemId)
              .put(File.Info.name)
              .put(File.Info.size);

        return holder.get();
    }

    private JSONObject itemInfoToJson(ItemInfo info) throws JSONException {
        JSONObjectHolder<ItemInfo> holder;

        holder = new JSONObjectHolder<ItemInfo>(info);
        holder.put(ItemInfo.itemId)
              .put(ItemInfo.name)
              .put(ItemInfo.type)
              .put(ItemInfo.size);

        return holder.get();
    }

    private JSONObject childrenToJson(Children info) throws JSONException {
        JSONObjectHolder<ItemInfo> holder;
        JSONArray array = new JSONArray();

        for(ItemInfo itemInfo: info.getContent()) {
            holder = new JSONObjectHolder<>(itemInfo);
            holder.put(ItemInfo.itemId)
                  .put(ItemInfo.name)
                  .put(ItemInfo.type)
                  .put(ItemInfo.size);

            try {
                array.put(holder.get());
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        JSONObject json = new JSONObject();
        json.put("items", array);
        return json;
    }

    private JSONObject hiveVoidToJson(Void padding) throws JSONException {
        JSONObjectHolder<Void> holder;

        holder = new JSONObjectHolder<Void>(padding);
        holder.put();

        return holder.get();
    }

    private JSONObject lengthToJson(Length length) throws JSONException {
        JSONObject json = new JSONObject();
        json.put("length", length.getLength());
        if (byteBuffer != null)
            json.put("data", new String(byteBuffer.array()));

        return json;
    }
}

File: src/ios/ResultHandler.swift
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import Foundation
import ElastosHiveSDK

class ResultHandler<T>: HiveCallback<T> {
    private var handlerId: Int
    private var callbackId: String
    private var commandDelegate: CDVCommandDelegate

    init(_ handlerId: Int, _ callbackId: String, _ commandDelegate:CDVCommandDelegate) {
        self.handlerId = handlerId
        self.callbackId = callbackId
        self.commandDelegate = commandDelegate
        super.init()
    }

    private func sendEvent(_ ret: Dictionary<String, Any>) {
        var dict = ret;
        dict["did"] = handlerId;
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: dict as [AnyHashable : Any]);
        result?.setKeepCallbackAs(true);
        self.commandDelegate.send(result, callbackId: self.callbackId);
    }

    public override func didSucceed(_ result: T) -> Void {
        var ret:Dictionary<String, Any>?

        if (result is HiveDriveHandle) {
            ret = driveHandleToDict(result as! HiveDriveHandle)
        } else if (result is HiveDirectoryHandle) {
            ret = directoryHandleToDict(result as! HiveDirectoryHandle)
        } else if (result is HiveFileHandle) {
            ret =  fileHandleToDict(result as! HiveFileHandle)
        } else if (result is HiveClientInfo) {
            ret =  clientInfoToDict(result as! HiveClientInfo)
        } else if (result is HiveDriveInfo) {
            ret =  driveInfoToDict(result as! HiveDriveInfo)
        } else if (result is HiveDirectoryInfo) {
            ret =  directoryInfoToDict(result as! HiveDirectoryInfo)
        } else if (result is HiveFileInfo) {
            ret =  fileInfoToDict(result as! HiveFileInfo)
        } else if (result is HiveItemInfo) {
            ret =  itemInfoToDict(result as! HiveItemInfo)
        } else if (result is HiveChildren) {
            ret =  childrenToDict(result as! HiveChildren)
        } else if (result is Void) {
            ret =  voidToDict(result as! Void)
        }
        sendEvent(ret!)
    }

    public override func runError(_ error: HiveError) -> Void {
        var ret = Dictionary<String, Any>()
        ret["error"] = error.localizedDescription
        sendEvent(ret)
    }

    private func driveHandleToDict(_ drive: HiveDriveHandle) -> Dictionary<String, Any> {
        let map = ObjectMap<HiveClientHandle>.acquire(DRIVE)
        let objId = DriveMap.nextObjId()

        ObjectMap.asDriveMap(map).put(objId, drive)

        var ret = Dictionary<String, Any>()
        ret["id"] = objId
        return ret
    }

    private func directoryHandleToDict(_ dir: HiveDirectoryHandle) -> Dictionary<String, Any> {
        let map = ObjectMap<HiveClientHandle>.acquire(DIR)
        let objId = DriveMap.nextObjId()

        ObjectMap.asDirMap(map).put(objId, dir)

        var ret = Dictionary<String, Any>()
        ret["id"] = objId
        return ret
    }

    private func fileHandleToDict(_ file: HiveFileHandle) -> Dictionary<String, Any> {
        let map = ObjectMap<HiveClientHandle>.acquire(FILE)
        let objId = FileMap.nextObjId()

        ObjectMap.asFileMap(map).put(objId, file)

        var ret = Dictionary<String, Any>()
        ret["id"] = objId
        return ret
    }

    private func clientInfoToDict(_ info: HiveClientInfo) -> Dictionary<String, Any> {
        let holder = DictionaryHolder<HiveClientInfo>(info)
        return holder.put(HiveClientInfo.userId)
            .put(HiveClientInfo.name)
            .put(HiveClientInfo.email)
            .put(HiveClientInfo.phoneNo)
            .put(HiveClientInfo.region)
            .get()
    }

    private func driveInfoToDict(_ info: HiveDriveInfo) -> Dictionary<String, Any> {
        let holder = DictionaryHolder<HiveDriveInfo>(info)
        return holder.put(HiveDriveInfo.driveId).get()
    }

    private func directoryInfoToDict(_ info: HiveDirectoryInfo) -> Dictionary<String, Any> {
        let holder = DictionaryHolder<HiveDirectoryInfo>(info)
        return holder.put(HiveDirectoryInfo.itemId)
            .put(HiveDirectoryInfo.name)
            .put(HiveDirectoryInfo.childCount)
            .get()
    }

    private func fileInfoToDict(_ info: HiveFileInfo) -> Dictionary<String, Any> {
        let holder = DictionaryHolder<HiveFileInfo>(info)
        return holder.put(HiveFileInfo.itemId)
            .put(HiveFileInfo.name)
            .put(HiveFileInfo.size)
            .get()
    }

    private func itemInfoToDict(_ result: HiveItemInfo) -> Dictionary<String, Any> {
        // TODO
        return Dictionary<String, Any>()
    }

    private func childrenToDict(_ result: HiveChildren) -> Dictionary<String, Any> {
        // TODO
        return Dictionary<String, Any>()
    }

    private func voidToDict(_ result: Void) -> Dictionary<String, Any> {
        // TODO
        return Dictionary<String, Any>()
    }
}
